File: frontend/src/services/apiError.ts
import { AxiosError } from "axios";
import type { ApiResponse, ApiError } from "../types/api";

export function extractApiErrorMessage(error: AxiosError): string {
  if (!error.response) {
    return "Unable to connect to server. Please try again!";
  }

  const data = error.response.data as ApiResponse<unknown>;
  const apiError: ApiError | undefined = data?.error;

  if (!apiError) {
    return data?.message || "An unknown error occurred.";
  }

  // If there are fieldErrors (validation errors)
  if (apiError.fieldErrors && apiError.fieldErrors.length > 0) {
    return apiError.fieldErrors
      .map((f) => `${f.field}: ${f.message}`)
      .join("\n");
  }

  // Prioritize friendly messages
  if (apiError.message) {
    return apiError.message;
  }

  // Fallback to debugMessage
  if (apiError.debugMessage) {
    return apiError.debugMessage;
  }

  // Final fallback
  return "An error occurred. Please try again!";
}

File: backend/src/main/java/com/example/backend/dto/response/ApiError.java
package com.example.backend.dto.response;

import java.util.List;

public record ApiError(
        String code,
        String message,
        String debugMessage,
        List<FieldError> fieldErrors
) {
    public record FieldError(String field, String code, String message) {}

    public static ApiError of(String code, String message) {
        return new ApiError(code, message, null, null);
    }

    public static ApiError of(String code, String message, String debugMessage) {
        return new ApiError(code, message, debugMessage, null);
    }

    public static ApiError validationError(List<FieldError> errors) {
        return new ApiError("VALIDATION_ERROR", "Invalid data", null, errors);
    }
}